src/qibolab/instruments/rfsoc.py

```diff
@@ -124,31 +143,32 @@ def convert_sweep(sweeper: Sweeper, sequence: PulseSequence, qubits: dict[int, Q
             raise ValueError("Sweeper amplitude is set to reach values higher than 1")
     else:
         for pulse in sweeper.pulses:
-            indexes.append(sequence.index(pulse))
+            idx_sweep = sequence.index(pulse)
+            indexes.append(idx_sweep)
             base_value = getattr(pulse, sweeper.parameter.name)
+            if idx_sweep != 0 and sweeper.parameter is START:
+                # do the conversion from start to delay
+                base_value = base_value - sequence[idx_sweep - 1].start
             values = sweeper.get_values(base_value)
             starts.append(values[0])
             stops.append(values[-1])
 
-            if sweeper.parameter not in {DURATION, START}:
-                parameters.append(convert_parameter(sweeper.parameter))
-            else:
-                if sweeper.parameter is DURATION:
-                    parameters.append(rfsoc.Parameter.DURATION)
-                elif sweeper.parameter is START:
-                    parameters.append(rfsoc.Parameter.START)
-
-                idx_sweep = sequence.index(pulse)
+            if sweeper.parameter is START:
+                parameters.append(rfsoc.Parameter.DELAY)
+            elif sweeper.parameter is DURATION:
+                parameters.append(rfsoc.Parameter.DURATION)
                 delta_start = values[0] - base_value
                 delta_stop = values[-1] - base_value
 
-                for next_pulse in sequence[idx_sweep + 1 :]:
-                    if next_pulse.qubit != pulse.qubit:
-                        continue
-                    parameters.append(rfsoc.Parameter.START)
-                    indexes.append(sequence.index(next_pulse))
-                    starts.append(next_pulse.start + delta_start)
-                    stops.append(next_pulse.start + delta_stop)
+                if len(sequence) > idx_sweep + 1:
+                    # if duration-swept pulse is not last
+                    indexes.append(idx_sweep + 1)
+                    t0 = sequence[idx_sweep + 1].start - sequence[idx_sweep].start
+                    parameters.append(rfsoc.Parameter.DELAY)
+                    starts.append(t0 + delta_start)
+                    stops.append(t0 + delta_stop)
+            else:
+                parameters.append(convert_parameter(sweeper.parameter))
```

Comment:
Just as a proposal for the future: I was thinking that we could make a single function `convert`, that determine the type on its own, and according to the type it will call the correct one (i.e. `convert_parameter`, `convert_sequence`, ...).

We can also give it the correct type with the `@overload` mechanism (this time there should be no generics around, so it should work smoothly).

Reply:
Yes that would probably be nicer but maybe also a bit more complex... Not sure.

Another thing that I was thinking was to add a `qibosoq.client` filw with the `_open_connection` function and with the `asdict` translations

Reply:
> Yes that would probably be nicer but maybe also a bit more complex... Not sure.

There is a bit more code, but I believe it would not be much more complex: you just keep the functions you currently have, and then add a single one:

```py
@overload
def convert(obj: Pulse) -> rfsoc.Pulse:
    ...

@overload
def convert(obj: Parameter) -> rfsoc.Parameter:
    ...

def convert(obj):
    if isinstance(obj, Pulse):
        return convert_pulse(obj)
    if isinstance(obj, Parameter):
        return convert_parameter(obj)
    ...
```

Eventually, we can also split all the conversion machinery to a separate file:
```
instruments/
  rfsoc/
    __init__.py  # here you import only the instrument class
    driver.py  # here you define the driver, but the conversion
    convert.py
```
Since the instrument class is the only thing required outside.

Reply:
> Another thing that I was thinking was to add a `qibosoq.client` filw with the `_open_connection` function and with the `asdict` translations

That might be a clever idea. However, the difference is that the client could also be implemented in Qibosoq, and potentially used in other places. The conversion instead is Qibolab specific, since you need to have access to both objects (*from* & *to*) and it's Qibolab that depends on Qibosoq, not the opposite.

Reply:
I will split the files in a different PR, one where I'll do just that